File: Python/Easy/trees/is_sub_tree.py
"""

Given the roots of two binary trees root and subRoot, 
return true if there is a subtree of root with the same 
structure and node values of subRoot and false otherwise.

A subtree of a binary tree tree is a tree that consists of a node in
tree and all of this node's descendants. 
The tree tree could also be considered as a subtree of 
itself.

"""

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def isSubtree(self, root, subRoot):
        def checkEqual(root, subroot):
            if root is None and subroot is None:
                return True
            elif root is None or subroot is None:
                return False
            elif root.val != subroot.val:
                return False
            else:
                return checkEqual(root.left, subroot.left) and checkEqual(root.right, subroot.right)
            
        if root is None:
            return False
        elif root == subRoot:
            return True
        else:
            return self.isSubtree(root.right, subRoot) or self.isSubtree(root.left, subRoot)

File: JavaScript/Easy/trees/is_subtree.js
/**
 * 
 * @param {TreeNode} root
 * @param {TreeNode} subRoot
 * @return {boolean}
 * 
 */

const TreeNode = (val, left, right) => {
    this.val = (val === undefined ? 0 : val)
    this.right = (right === undefined ? null : right)
    this.left = (left === undefined ? null : left)
}

var isSubTree = function(root, subRoot) {
    // Write helper function to check if equal (same as simpler tree problems done previously)
    var checkEqual = (root1, root2) => {
        if (root1 === null && root2 === null) {return true}
        else if (root1 === null || root2 === null) {return false}
        else if (root1.val !== root2.val) {return false}
        else {return checkEqual(root1.left, root2.left) && checkEqual(root1.right, root2.right)}
    }

    if (root === null) {return false}
    else if (checkEqual(root, subRoot)) {return true}
    else {return isSubTree(root.right, subRoot) || isSubTree(root.left, subRoot)}

}